source/server-selection/server-selection.rst

```diff
@@ -1705,17 +1732,42 @@ The user's intent in specifying two tag sets was to fall back to the second set
 if needed, so we filter by maxStalenessSeconds first, then tag_sets, and select
 Node 2.
 
+Why change from pure random selection when selecting from within the latency window?
+------------------------------------------------------------------------------------
+
+When a node slows down, pooled connections to it will remain checked out for
+longer periods of time due to operations taking longer to complete
+server-side. Assuming at least constant incoming operation load, more
+connections will need to be opened against the node to service new operations,
+further straining it and slowing it down. This can lead to runaway connection
+creation scenarios that can cripple a deployment ("connnection storms"). As part
+of DRIVERS-781, the load balancing algorithm was changed to more evenly spread
+out the workload among suitable servers to prevent any single node from being
+overloaded. The new algorithm achieves this by routing operations to servers
+with fewer active connections, thereby reducing the number of new operations and
+connection creations routed towards nodes that are busier. The previous random
+selection mechanism did not take load into account and could assign work to
+nodes that were under too much stress already.
+
+As an added benefit, the new algorithm gives preference to nodes that have
+recently been discovered and are thus are more likely to be alive (e.g. during a
+rolling restart). The narrowing to two random choices first ensures new servers
+aren't overly preferred however, preventing a "thundering herd" situation.
```

Comment:
Doesn't this algorithm suffer from the "thundering herd" problem as soon as the "servers within the latency window" goes down to 2 servers? In the pure random selection each node would get 50% of the work, whereas in the new algorithm the new server would (potentially) get 100% of the work.

We could also consider what happens when the "servers within the latency window" goes down to 3 servers. In pure random selection each node would get 33% of the work, whereas in the new algorithm the new server would (potentially) get 66% of the work.

With 4 servers, pure random gives 25% whereas the new algorithm gives a 50% worst case.


Is this acceptable? Should we consider only using this power of two choices algorithm when the number of servers in the latency windows is above some threshold (maybe 3)?

Reply:
I think in cases where there are only a few available nodes, the "thundering herd" problem is less of a concern. For example, in the case where we go from 1->2 servers, it's not really any more of a "herd" than was already happening when we were routing every single operation to the lone node. We could have an issue in a situation where the lone node is isolated but the new node is shared among driver instances, but I'm not sure if this is a case that actually happens in practice. Also remember that targeting newer nodes is a benefit of this algorithm, so imo it's okay than in the 3 node and 4 node cases that the newer server is seeing some higher load initially.

Reply:
I agree in principle with the idea of routing operations to new servers. When I said "it's not very straightforward to determine why this statement is true" I meant that it's difficult to understand how (and to what degree) this new algorithm satisfies that goal. I hope the new test format you added will make it easier to reason about the behavior. 

> On the flip side of that example, while A is marked as unknown after the transient network error, B would be accepting 100% of the requests, so ideally we'd want to prefer the new server as soon as possible to start shedding some of that load.

But in this case we wouldn't be shedding _some_ of the load. We would be shedding _all_ the load. I think the scenario above is certainly a potential regression where request load can unnecessarily flip-flop between two servers. I wonder if this case should be mentioned in the spec. It might prove useful down the line if we see a similar issue in the field.

Reply:
I don't think the new test format will necessarily help shed too much light on how the algorithm performs in practice since the tests are more of a point-in-time verification. A key benefit of the algorithm in practice is that routing operations to an individual node makes it less likely for that node to receive future operations until the routed operations complete, but that sort of thing isn't accounted for in the spec tests, which perform the same selection many times. I'm going to do some ad-hoc experimentation on real deployments with the Rust driver to see how the algorithm actually handles a few different scenarios (this one included), and I'll let you know the results of that.

>But in this case we wouldn't be shedding some of the load. We would be shedding all the load. I think the scenario above is certainly a potential regression where request load can unnecessarily flip-flop between two servers.

This is true, but only for short periods of time due to the convergence of activeConnectionCounts I alluded to in my previous comment. In fact, both nodes should reach equilibrium points more quickly under this scheme than in pure random chance.

Sending too much load to an individual node (hotspotting) even for a short period of time is certainly a concern though, but I think it's mostly kept in check by maxConnecting. For instance, current drivers would route roughly 50% of the load to the new server once it comes back up, but they will do so via a burst of new connections. In the new scheme, drivers will temporarily route 100% of the load to the new node, but only 2 at a time until the pool starts to fill. So it might be that, even though drivers will be routing a higher % of operations to the node once it comes up, the new node will actually receive a more manageable workload with this algorithm + maxConnecting than it would with existing drivers. There is a risk that overall latency worsens, though. I'll have to see when I do my experiments.

This does make me realize that we might need to take into account the length of the WaitQueue when calculating load or health of a server. In this scenario for instance, a bunch of operations could be queued on the new node at the start up due to maxConnecting, but activeConnectionCount would still be 2, leading to further operations getting routed to the node in the meantime. If we factored in the WaitQueue size to the comparisons we do, we'd potentially start routing operations to the other node sooner. We also get the added benefit of evenly distributing queued work when all connection pools are full. It's hard for me to know how much of a problem this is though. What do you think?

Reply:
So I performed a few experiments on a sharded cluster with 2 mongoses. In each experiment, I spawned 50 tasks that each ran `find_one` a hundred times, and I measured how long it took to finish, how many times each shard was selected, and the connections created in each pool to service this workload. Besides one scenario noted below, all pools began as empty. I performed this experiment with the following scenarios:
- 2 normal mongoses
- 2 mongoses, 1 of which (A) has a 5% chance of `find` taking an extra 250ms
- 2 mongoses, 1 of which (A) has a 25% chance of `find` taking an extra 250ms
- 2 mongoses with minPoolSize=100, 1 of which (A) has a 25% chance of `find` taking an extra 250ms
    - I waited until the pools both reached 100 connections before starting this scenario
- 2 mongoses, both have a 5% chance of `find` taking an extra 250ms
- 2 mongoses, both have a 25% chance of `find` taking an extra 250ms

Here are the results: https://docs.google.com/spreadsheets/d/1r9CZRjPEbk43g5BLQFXY0Ql4iTlzw2qqyt2dj583XRs/edit?usp=sharing

Some notable points:
- Pure random appears slightly faster in "normal" scenarios (i.e. 1, 2, 5). This seems to be due to the fact that considering health spreads connection creation out more evenly, resulting in more connections being created and more latency. The differences were really narrow though and this benchmarking was certainly not performed in an ideal environment, so its not clear if this difference is meaningful or not.
- Considering health leads to the driver preferring better performing nodes.
- Surprisingly, considering health performed worse (latency-wise) than pure random in scenario 3. This also seems to be due to considering health creating more connections in total, which seems to outweigh the latency benefits of routing away from the slow node. This hypothesis is potentially confirmed by scenario 4, which populates the pools ahead of time and has health performing significantly better than pure random. Given a much larger operation count, we'd likely see health perform better in scenario 3 as well.
- Considering health leads to more evenly populated pools.
- Considering health performed better than pure random in Scenario 6 (lots of ops on both mongoses take a long time). This seems to align with the intended goal of better handling of slow operations.

None of these workloads actually made the pools reach maxPoolSize and start queuing operations, though when I did try one that did (e.g. burst of 5000 tasks at once), the results started to look the same between the two methods. This seems to indicate that we do in fact need to consider the length of the WaitQueue when calculating load, since otherwise both pools will be considered under equal load when they're full and we'll just choose at random. I'll update the spec to include this. 

Note that I omitted the scenario we've been discussing here. I still need to figure out a good way to emulate that and record proper results. I figured these preliminary ones were worth sharing though.